pennylane/devices/qubit/simulate.py

```diff
@@ -57,7 +58,7 @@ def simulate(circuit: qml.tape.QuantumScript, rng=None) -> Result:
     state = create_initial_state(circuit.wires, circuit._prep[0] if circuit._prep else None)
 
     for op in circuit._ops:
-        state = apply_operation(op, state)
+        state = apply_operation(op, state, debugger=debugger)
```

Comment:
do we prefer adding the `debugger` kwarg to every dispatched `apply_operation` over adding something hard-coded here, like:
```python
if isinstance(op, qml.Snapshot):
    if debugger is not None:
        # the stuff
    continue
state = apply_operation(op, state)
```
this comes at the cost of an extra isinstance check (since singledispatch will do it again assuming it's not a Snapshot)

Reply:
I think we have to trade between the complexity of the call signature for `apply_operation` and the complexity of the logical structure in `simulate`. 

In the future, I think we should consider different ways of doing snapshots, such as making it so that the snapshot operation itself points to where the state can get stored.  That would eliminate the need to have a debugger object on the device that gets passed around.

But, this opens up the behaviour.  I don't think we need to spend that much time on something like this right now.